applications: Asset Tracker v2: Add debug module that uses Memfault

simensrostad opened the pull request:
Add new debug module that subscribes to events from all modules in the
system. The module integrates basic support for Memfault with
custom GPS metrics in addition to the pre-existing metrics already
supported in NCS.

Closes CIA-355 and CIA-356

balaji-nordic commented on applications/asset_tracker_v2/src/events/Kconfig:
```diff
@@ -42,6 +42,10 @@ config DATA_EVENTS_LOG
 	default y
 
+config DEBUG_EVENTS_LOG
+	bool "Enable logging for Cloud module events"
```
Cloud->Debug

balaji-nordic commented on applications/asset_tracker_v2/src/events/debug_module_event.h:
```diff
@@ -0,0 +1,57 @@
+	size_t len;
+};
+
+/** @brief Debug event. */
```
I dont think this short @brief is needed.

simensrostad replied:
Maybe not, I includeded it to align with the other module event docs in the application.

balaji-nordic commented on applications/asset_tracker_v2/src/modules/Kconfig.cloud_module:
```diff
@@ -12,8 +12,8 @@ if CLOUD_MODULE
 	int "Cloud module thread stack size"
-	default 4096 if NRF_CLOUD
-	default 2688
+	default 6144 if NRF_CLOUD_MQTT
```
Just out of curiousity: What did we change NRF_CLOUD to NRF_CLOUD_MQTT in this PR? And why did the default stack size increase? If this is only because of debug module, then shouldn't we increase the thread size only if the debug module is configured?

simensrostad replied:
After https://github.com/nrfconnect/sdk-nrf/pull/4609. Kconfigs associated with `NRF_CLOUD`  were split into separate module Kconfigs. `CLOUD_THREAD_STACK_SIZE` is now ultimately dependent on `NRF_CLOUD_MQTT` and not `NRF_CLOUD`. `NRF_CLOUD` is now more of a umbrella config for everything related to nRF Cloud libraries. We only care about the MQTT part in this case.

You`re right! The stack increase should depend on `CONFIG_DEBUG_MODULE_USE_CUSTOM_TRANSPORT` being enabled.

balaji-nordic commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,47 @@
+if DEBUG_MODULE
+
+config DEBUG_MODULE_USE_CUSTOM_TRANSPORT
+	bool "Use custom transport when sending memfault data"
```
Should we not have a "default" value here?

simensrostad replied:
Kconfig options automatically defaults to "n". So adding `default n` is not needed.

balaji-nordic commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,47 @@
+config DEBUG_MODULE_USE_CUSTOM_TRANSPORT
+	bool "Use custom transport when sending memfault data"
+	help
+	  If this option is disabled, Memfault firwmare SDKs own internal HTTP transport is used to
```
```suggestion
         If enabled, the debug module will forward the data in chunks via
	  events of type DEBUG_EVT_MEMFAULT_DATA_READY via the
	  configured cloud to Memfault. If disabled, the debug module will use the 
	  Memfault firmware SDKs internal HTTP transport. This can be used to save overhead 
	  associated with maintaining multiple connections.```

balaji-nordic commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,47 @@
+config DEBUG_MODULE_USE_CUSTOM_TRANSPORT
+	bool "Use custom transport when sending memfault data"
+	help
+	  If this option is disabled, Memfault firwmare SDKs own internal HTTP transport is used to
```
Propose to word it this way.

          If enabled, the debug module will forward the data in chunks using
	  the events of type DEBUG_EVT_MEMFAULT_DATA_READY via the
	  configured cloud to Memfault. If disabled, the debug module will use the 
	  Memfault firmware SDKs internal HTTP transport. This can be used to save overhead 
	  associated with maintaining multiple connections.```

balaji-nordic replied:
Also would'nt it always transmit "chunks" irrespective of the transport? If thats true, then you can just drop the wording about 'chunks'.

simensrostad replied:
Will update the description and remove the working around `chunks`.

balaji-nordic commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,47 @@
+	  Configuration that sets how often Memfault captures and prepares data that will be sent
+	  to Memfault cloud. Reduce this interval to get more frequent updates to Memfault.
+	  Note that data transmitted to Memfault cloud is still dependent on how often the
+	  application sends data, which can be manipulated by setting the applications real-time
```
```suggestion
	  application sends data, which can be manipulated by setting the application's real-time
```

balaji-nordic replied:
What are those 'real-time configurations'?

simensrostad replied:
The applications real-time configurations are the configurations that governs the overall behavior of the device. They are documented in the root README.rst

balaji-nordic commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,47 @@
+
+config DEBUG_MODULE_MEMFAULT_CHUNK_SIZE_MAX
+	int "Fragment size of data transmitted to memfault"
+	default 2048 if DEBUG_MODULE_USE_CUSTOM_TRANSPORT
```
Shoult it be so high when using custom transport?

simensrostad replied:
I'm afraid it has to. Or else the cloud module will get spammed with tens of events of type `DEBUG_EVT_MEMFAULT_DATA_READY` in case of a core-dump, carrying smaller amounts of data. Increasing this option makes the debug module send less events carrying more data. This way the internal processing queue in the cloud module won't get overfilled. (if that happens we must to a reboot to avoid undefined behavior in the app),.

balaji-nordic commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,47 @@
+	default 2048 if DEBUG_MODULE_USE_CUSTOM_TRANSPORT
+	default 80
+	help
+	  Decreasing this option will in case of a core-dump produce more events carrying
```
```suggestion
	  Decreasing this option will, in case of a core-dump, produce more events carrying
```

balaji-nordic commented on applications/asset_tracker_v2/src/modules/cloud_module.c:
```diff
@@ -392,6 +401,19 @@ static void data_send(struct data_module_event *evt)
+
+	err = cloud_wrap_memfault_data_send(evt->data.memfault.buf, evt->data.memfault.len);
+	if (err) {
+		LOG_ERR("cloud_wrap_data_send, err: %d", err);
```
```suggestion
		LOG_ERR("cloud_wrap_memfault_data_send, err: %d", err);
```

balaji-nordic commented on applications/asset_tracker_v2/src/modules/debug_module.c:
```diff
@@ -0,0 +1,288 @@
+
+#include <zephyr.h>
+#include <logging/log.h>
+#if defined(CONFIG_MEMFAULT)
```
defined(CONFIG_MEMFAULT) -> !defined(CONFIG_DEBUG_MODULE_USE_CUSTOM_TRANSPORT) ?

I am assuimg !DEBUG_MODULE_USE_CUSTOM_TRANSPORT implies CONFIG_MEMFAULT

simensrostad replied:
Actually `CONFIG_DEBUG_MODULE_USE_CUSTOM_TRANSPORT` depends on `CONFIG_MEMFAULT`. `CONFIG_DEBUG_MODULE_USE_CUSTOM_TRANSPORT` selects what transport Memfault data will be sent on. Ill update the Kconfig name to make this more clear

simensrostad replied:
Ill update it to `DEBUG_MODULE_MEMFAULT_USE_EXTERNAL_TRANSPORT` if no one disagrees.

balaji-nordic replied:
Seems a bit wordy. But I cant think of any other better alternative. So go for it.

balaji-nordic commented on applications/asset_tracker_v2/src/modules/debug_module.c:
```diff
@@ -0,0 +1,288 @@
+}
+
+EVENT_LISTENER(MODULE, event_handler);
+EVENT_SUBSCRIBE_EARLY(MODULE, app_module_event);
```
Why does the debug module have to be an early subscriber to all the events? I was assuming that this be the late subscriber because the data processed and generated by this module is not delay sensitive.

simensrostad replied:
In some cases, modules consume events. Meaning that they are not carried around to other modules after the respective module that consumed the event has processed it. Subscribing early is just a precaution to ensure that the debug module most likely receives all events before they are potentially discarded.

balaji-nordic commented on applications/asset_tracker_v2/src/modules/gps_module.c:
```diff
@@ -52,6 +52,15 @@ static enum sub_state_type {
+	uint64_t start_uptime;
+	/* Numer of satellites tracked for a GPS fix or timeout. */
+	uint8_t satellites_tracked;
+} helpers;
```
Why the name "helpers"? Shouldn't it rather be "stats" or "statistics"?

simensrostad replied:
No opinion here really. I can rename it to `stats`

balaji-nordic commented on applications/asset_tracker_v2/src/modules/gps_module.c:
```diff
@@ -52,6 +52,15 @@ static enum sub_state_type {
+	 * GPS fix and timeout.
+	 */
+	uint64_t start_uptime;
+	/* Numer of satellites tracked for a GPS fix or timeout. */
```
```suggestion
	/* Number of satellites tracked before getting a GPS fix or a timeout. */
```

simensrostad commented on applications/asset_tracker_v2/src/modules/debug_module.c:
```diff
@@ -0,0 +1,288 @@
+		 * timeout. This is to catch a timeout in advance so that Memfault can save
+		 * coredump data before a reboot occurs.
+		 */
+		err = memfault_software_watchdog_update_timeout(evt->timeout - 5000);
```
Good idea. Ill add an assert.

simensrostad commented:
@umapraseeda Could you have a look at the documentation?

simensrostad commented on applications/asset_tracker_v2/src/modules/gps_module.c:
```diff
@@ -52,6 +52,15 @@ static enum sub_state_type {
+	 * GPS fix and timeout.
+	 */
+	uint64_t start_uptime;
+	/* Number of satellites tracked before getting a GPS fix or a timeout. */
```
It should have been changed to your suggestion?

balaji-nordic replied:
You are right. This is fixed. For some reason, I did not see it.

simensrostad commented on applications/asset_tracker_v2/src/modules/debug_module.c:
```diff
@@ -0,0 +1,299 @@
+/* Time before the actual watchdog timeout that the Memfault software watchdog timeout will trigger.
+ * In milliseconds.
+ */
+#define MEMFAULT_SOFTWARE_WATCHDOG_DELTA 5000
```
Jepp. Ill add a Kconfig

simensrostad commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,49 @@
+
+config DEBUG_MODULE_MEMFAULT_HEARTBEAT_INTERVAL_SEC
+	int "Memfault heartbeat interval in seconds"
+	default 1800
```
Ill go through the Kconfigs and update potential dependencies. I'm also having some issue with the Unity test when depending on MEMFAULT.

jtguggedal commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,48 @@
+	  application sends data, which can be manipulated by setting the application's real-time
+	  configurations.
+
+endif
```
This `endif` should probably be at the end of the file

simensrostad replied:
I removed the if guard for these statements as `DEBUG_MODULE_MEMFAULT_CHUNK_SIZE_MAX` is used in `debug_module_event.h` which is included in the `cloud_module`. Either we define a fixed size buffer in `debug_module_memfault_data` (which must be rather big) or we need to have `DEBUG_MODULE_MEMFAULT_CHUNK_SIZE_MAX` always exposed.

jtguggedal commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,48 @@
+	  cloud module message queue depending on how fast the cloud module thread can handle
+	  incoming queue messages. An overfilled message queue is in general considered a fatal
+	  error that can cause undefined behaviour. To reduce the likelyhood of this issue
+	  occurring, this value should be kept reasonably high.
```
`, but within the modem firmware's documented limitations`

jtguggedal commented on applications/asset_tracker_v2/src/modules/cloud_module.c:
```diff
@@ -91,7 +93,7 @@ const k_tid_t cloud_module_thread;
 
 /* Cloud module message queue. */
-#define CLOUD_QUEUE_ENTRY_COUNT		10
+#define CLOUD_QUEUE_ENTRY_COUNT		20
```
We should Kconfigify this configuration

simensrostad replied:
Agreed. Its in the backlog to do this for all modules in the application.

jtguggedal commented on applications/asset_tracker_v2/src/modules/debug_module.c:
```diff
@@ -0,0 +1,299 @@
+	bool data_available = memfault_packetizer_data_available();
+
+	if (IS_ENABLED(CONFIG_DEBUG_MODULE_MEMFAULT_USE_EXTERNAL_TRANSPORT) && data_available) {
+
```
redundant blank line

balaji-nordic commented on applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:
```diff
@@ -0,0 +1,236 @@
+{
+	setup_debug_module_in_init_state();
+
+	/* Update this function to expect the search time and number of satellites. */
```
Unwanted comment?

balaji-nordic commented on applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:
```diff
@@ -0,0 +1,236 @@
+						MEMFAULT_METRICS_KEY(GpsSatellitesTracked),
+						2,
+						0);
+	__wrap_memfault_metrics_heartbeat_debug_trigger_Ignore();
```
Shouldn't this be an Expect than an Ignore?

simensrostad replied:
Jepp.

balaji-nordic commented on applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:
```diff
@@ -0,0 +1,236 @@
+	k_free(app_module_event);
+}
+
+void test_memfault_trigger_metric_sampling_on_gps_fix(void)
```
Can you write one or two lines about what each test case does?

simensrostad commented on applications/asset_tracker_v2/overlay-debug.conf:
```diff
@@ -64,3 +64,19 @@ CONFIG_NRF9160_GPS_LOG_LEVEL_DBG=y
+# CONFIG_MEMFAULT=y
+# CONFIG_MEMFAULT_NCS_PROJECT_KEY=""
+# CONFIG_MEMFAULT_NCS_LTE_METRICS=y
+# CONFIG_MEMFAULT_NCS_STACK_METRICS=y
```
Agreed, lets to that.

balaji-nordic commented on applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:
```diff
@@ -0,0 +1,236 @@
+
+	struct gps_module_event *gps_module_event = new_gps_module_event();
+
+	gps_module_event->type = GPS_EVT_ACTIVE;
```
Can you extend this test to send all unhandled events? (If thats cumbersome, please change the name of this test to test_memfault_no_memfault_call_on_gps_evt_active())

simensrostad replied:
Jepp. will add test for all other unhandled events.

simensrostad commented on applications/asset_tracker_v2/src/cloud/aws_iot_integration.c:
```diff
@@ -32,6 +32,8 @@ LOG_MODULE_REGISTER(MODULE, CONFIG_CLOUD_INTEGRATION_LOG_LEVEL);
 #define PGPS_REQUEST_TOPIC_LEN (AWS_CLOUD_CLIENT_ID_LEN + 9)
 #define PGPS_RESPONSE_TOPIC "%s/pgps"
 #define PGPS_RESPONSE_TOPIC_LEN (AWS_CLOUD_CLIENT_ID_LEN + 5)
+#define MEMFAULT_TOPIC "%s/memfault"
```
Yes. Lets discuss this further when the cloud side is implemented.

simensrostad commented on applications/asset_tracker_v2/prj.conf:
```diff
@@ -15,7 +15,7 @@ CONFIG_FPU=y
 CONFIG_HEAP_MEM_POOL_SIZE=32768
 CONFIG_MAIN_STACK_SIZE=1280
-CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=3584
```
its only necessary to increase this option if `DEBUG_MODULE_MEMFAULT_USE_EXTERNAL_TRANSPORT` if enabled. Moved the config to `overlay-memfault.conf`

simensrostad commented on applications/asset_tracker_v2/configuration/memfault/memfault_metrics_heartbeat_config.def:
```diff
@@ -0,0 +1,7 @@
+/* Application-specific metrics can be defined here.
+ * Please refer to https://docs.memfault.com/docs/embedded/metrics-api for more details.
```
I like that it contains the links to the Memfault docs. But I don't have any strong opinion on this. I can remove it if you think its unnecessary.

jtguggedal replied:
Yeah, it has some value to have the link. It was mostly the ` can be defined here` part I found a bit misleading as there's already some metrics here. Maybe that can be cut, so that it reads `Application-specific metrics`, but it's not a bug deal. Don't push and re-run CI just for this change

simensrostad commented:
@jtguggedal @balaji-nordic _I think_ I've addressed all your review comments. @umapraseeda The documentation added in this PR should be ready for review.

jtguggedal commented on applications/asset_tracker_v2/src/modules/Kconfig.debug_module:
```diff
@@ -0,0 +1,57 @@
+
+menuconfig DEBUG_MODULE
+	bool "Debug module"
+
```
I think we need a guard here somewhere, otherwise all these options will show up in menuconfig and other UIs also when the debug module is disabled

balaji-nordic reviewed (approved):
Looks good now. 👍

jtguggedal reviewed (approved):
LGTM, thanks!
